### PR TITLE
add ExecutionConfig class

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
A first step as part of the execution pipeline rework.

**Description of the Change:**
Adds an `ExecutionConfig` class with some default values using the dataclass decorator for convenience. Also created the `workflow` module to encapsulate all execution pipeline work.

**Benefits:**
This will serve as a building block and allow us to move forward in the rework.

### CODE REVIEW

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:29]: I learned recently about `typing.Optional`. Instead of `Union[None, int]`, we can use `Optional[int]`.

Additional question, should we store shot vectors in the `shots` property or in an additional `shot_vector` property? If so, shots could also be `Tuple[int]`.

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:32]: How do we want to represent backpropagation and device gradients here?

File: pennylane/workflow/execution.py
Comment (albi3ro) [inline pennylane/workflow/execution.py:35]: I'm a little unsure where these parameters would be coming from and where they'd be getting used. Can you give me an example?

File: pennylane/workflow/execution.py
Comment (timmysilv) [reply pennylane/workflow/execution.py:29]: I originally had shots as a union with tuple, but I didn't really see any previous examples for `shots` like this. I prefer it though, so I'll put it back. `Optional[Union[int, Tuple[int]]]` lol

File: pennylane/workflow/execution.py
Comment (timmysilv) [reply pennylane/workflow/execution.py:32]: this is the same as `diff_method` right? I'm good to have this copy the `diff_method` inputs if it makes more sense, but I don't love receiving strings when we can instead receive something like an enum, but those can be clumsy.

File: pennylane/workflow/execution.py
Comment (timmysilv) [reply pennylane/workflow/execution.py:35]: idk. the story said gradient hyperparameters should be configurable here, so I added them :p

File: pennylane/workflow/execution.py
Comment (albi3ro) [reply pennylane/workflow/execution.py:32]: Both `diff_method` provided to the QNode and `gradient_fn` provided to `execute` can be a string or a callable. Basically, the documentation for `qml.execute` is wrong, because `gradient_fn` can be either "device" or "backprop".

File: pennylane/workflow/execution.py
Comment (albi3ro) [reply pennylane/workflow/execution.py:35]: Ah. So what I'm meant is [these things](https://github.com/PennyLaneAI/pennylane/blob/359130c783b48fea04bac38c5602df2cc3aa401d/pennylane/gradients/gradient_transform.py#L22). In QNode, they are handled fairly suboptimally as the `gradient_kwargs`.

Maybe just `gradient_hyperparameters` instead? With the description that they are used in by the gradient transform.

File: doc/index.rst
Comment (albi3ro) [inline doc/index.rst:194]: Should we stick this under "Internals" for now? We can move it up to "API" later.

**Review by albi3ro — APPROVED** [review]
Great place to get us started.  We can continue to iterate on this throughout the quarter as we start to integrate it.

This should be a two-approval PR, as its a new class.

**Comment by timmysilv** [comment]
I'd like to note. The `dataclasses` package actually has a `field` function to accomplish the empty default dict in the manner that I'm doing using `default_factory=dict`. Unfortunately, using the `field` function makes that field not appear on the sphinx doc as an attribute (sphinx-doc/sphinx#10893), so I'm setting those manually in the `__post_init__` hook.

File: doc/code/qml_workflow.rst
Comment (rmoyard) [inline doc/code/qml_workflow.rst:1]: I would avoid creating a module without a plan shared and accepted by the team and the product team. Without it I am afraid that we will have to deprecate it in some months after we realize that it is not useful or does not fit our needs. In the meantime, I would add that in the interface module where `execution.py` is.

File: pennylane/workflow/execution.py
Comment (rmoyard) [inline pennylane/workflow/execution.py:27]: We will need to add spsa after https://github.com/PennyLaneAI/pennylane/pull/3440

**Review by rmoyard — COMMENTED** [review]
Hey @timmysilv, it looks great :100: Some comments: 

1. I don't think we should create the workflow module before having a strict plan for it. I think we should add that after having a good understanding on how the new devices will affect the execution pipeline and do the changes when we have a pipeline that can be run to add.
2. About using framework instead of interface will introduce a breaking change at some point and I want to be sure that the product team agrees with it before moving forward.

File: doc/code/qml_workflow.rst
Comment (timmysilv) [reply doc/code/qml_workflow.rst:1]: I agree that this `workflow` module is coming out of nowhere, but I don't feel like `interfaces` is also not the best home (since we are trying to isolate new code from old code). While I am not certain that this is where it will end up, this certainly enables us developers now while maintaining the separation needed to get this into the desired state once more planning/discussion has happened.

Perhaps we should remove the documentation until we have more certainty about how everything will end up? `ExecutionConfig` isn't imported at the top level as of this PR.

File: pennylane/workflow/execution.py
Comment (timmysilv) [reply pennylane/workflow/execution.py:27]: I'd like for this list to ultimately exist in only one place after this overhaul

File: pennylane/workflow/execution.py
Comment (timmysilv) [reply pennylane/workflow/execution.py:53]: I'm not sure of the product team's awareness of this, but it was discussed earlier this week in some meeting. I feel like it's a good move, but I'm happy to stick to interface if we feel that it's necessary.

I made the personal decision of avoiding `ml_framework` because I thought it looked rather awkward, but I see both you and @albi3ro proposed it so I will change this (and my other) PR to use `ml_framework`.

File: doc/code/qml_workflow.rst
Comment (albi3ro) [reply doc/code/qml_workflow.rst:1]: Simple solution for now: Just don't build it on the website.

That will keep this an internal implementation detail until the module is better fleshed out. Source code locations are purely implementation details; import paths for user-facing objects are product details.

As this object's only "interface" connection is a single property called "framework" and it will interact with qnodes, quantum scripts, and devices, it does not belong in `interfaces` at all.

File: pennylane/workflow/execution.py
Comment (josh146) [reply pennylane/workflow/execution.py:53]: while this is just an internal implementation for now, I would err on simply keeping `interface` for now pending a decision on porting everything. Partly to avoid confusion, and partly to avoid a situation where we end up with _two_ 'public' conventions.

Comment (josh146) [reply pennylane/workflow/execution.py:53]: But will defer to @trbromley here

File: doc/code/qml_workflow.rst
Comment (rmoyard) [reply doc/code/qml_workflow.rst]: I don't think it is an implementation detail when we create a new module even if we do not add documentation. We already have two execute pipelines, I am not in favour to add a third partially working one. This project (the module) is a great occasion to clean and create a better pipeline and we should plan accordingly and discuss the main points. I don't think the planning can be done until we have merged a working device. It should be postpone until:

1. The new device work is in a good state and unit tests are added (avoid relying on integration tests for coverage).
2. The workflow module and a new pipeline are approved by @josh146 and @trbromley 

Or all the workflow PRs are made on a separate branch `worflow` that is not merged in master until we have a satisfying new execution pipeline.

Comment (timmysilv) [reply doc/code/qml_workflow.rst]: fair enough. if we intend on _not_ having something complete/functional where we maintain a WIP pipeline, I'd prefer to do this work on a `workflow` branch

**Review by rmoyard — APPROVED** [review]
Thanks @timmysilv

**Pull request merged**
